src/core/utils/flash.cpp

```diff
@@ -225,71 +522,97 @@ bool Flash::DoesValidRecordExist(uint32_t aOffset, uint16_t aKey) const
 
 void Flash::Swap(void)
 {
-    uint8_t  dstIndex  = !mSwapIndex;
-    uint32_t dstOffset = kSwapMarkerSize;
-    Record   record;
+    SwapHeader swapHeader;
+    uint8_t    dstIndex  = !mSwapIndex;
+    uint32_t   dstOffset = sizeof(swapHeader);
+    Record     record;
+    uint16_t   size;
+    uint16_t   eraseCounter = GetEraseCounter();
 
     otPlatFlashErase(&GetInstance(), dstIndex);
 
-    for (uint32_t srcOffset = kSwapMarkerSize; srcOffset < mSwapUsed; srcOffset += record.GetSize())
+    if (dstIndex == 0 && eraseCounter < 0xffff)
+    {
+        eraseCounter++;
+    }
+
+    for (uint32_t srcOffset = mSwapHeaderSize; srcOffset < mSwapUsed; srcOffset += size)
     {
-        otPlatFlashRead(&GetInstance(), mSwapIndex, srcOffset, &record, sizeof(RecordHeader));
+        record.LoadHeader(&GetInstance(), mSwapIndex, srcOffset);
 
-        VerifyOrExit(record.IsAddBeginSet(), OT_NOOP);
+        VerifyOrExit(!record.IsFree(), OT_NOOP);
 
-        if (!record.IsValid() || DoesValidRecordExist(srcOffset + record.GetSize(), record.GetKey()))
+        size = record.GetSize();
+
+        if (!record.IsValid() || IsInvalidated(srcOffset + record.GetSize(), record.GetKey()))
         {
             continue;
         }
 
-        otPlatFlashRead(&GetInstance(), mSwapIndex, srcOffset, &record, record.GetSize());
-        otPlatFlashWrite(&GetInstance(), dstIndex, dstOffset, &record, record.GetSize());
+        record.LoadData(&GetInstance(), mSwapIndex, srcOffset);
+#if OPENTHREAD_CONFIG_FLASH_LEGACY_COMPATIBILITY
+        record.Fixup();
+#endif
+        record.Save(&GetInstance(), dstIndex, dstOffset);
```

Comment:
Should we validate integrity of record after writing? If integrity check fails, we can try append a new record?